In [1]:
# Start Time
from datetime import datetime
start_time1 = datetime.now()

In [2]:
# importing dependencies
import pandas as pd
import math

In [3]:
# Importing Source Sheet
df_src = pd.read_excel('Problem_Dataset.xlsx' , sheet_name=0 )

In [4]:
# Importing Destination Sheet
df_des = pd.read_excel('Problem_Dataset.xlsx' , sheet_name=1)

In [5]:
df_src.head()  # top 5 records

Latitude  Longitude
0  13.022193  77.697762
1  13.030745  77.558343
2  12.928784  77.509375
3  13.026320  77.720210
4  12.948364  77.627783

In [6]:
df_src.shape  

(5175, 2)

In [7]:
df_des.head()  # top 5 records

Latitude  Longitude
0  28.519427  76.964814
1  28.521104  76.996469
2  28.522902  76.910281
3  28.523170  76.986546
4  28.526302  76.953613

In [8]:
df_des.shape 

(4817, 2)

In [9]:
df_src = df_src.iloc[0:: , ::]  # selecting all records from Source

In [10]:
df_des = df_des.iloc[0:: , ::]  # selecting all records from destination

### Source Data

In [11]:
# We create here new column for SOURCE_ID values
df_src.insert(loc=0, column='SOURCE_ID', value=range(0 , 0 + len(df_src)))

In [12]:
df_src.rename(columns={'Latitude' : 'SOURCE_Latitude' , 'Longitude' : 'SOURCE_Longitude'} , inplace=True)

In [13]:
df_src

SOURCE_ID  SOURCE_Latitude  SOURCE_Longitude
0             0        13.022193         77.697762
1             1        13.030745         77.558343
2             2        12.928784         77.509375
3             3        13.026320         77.720210
4             4        12.948364         77.627783
...         ...              ...               ...
5170       5170        13.012562         77.605651
5171       5171        13.003766         77.612212
5172       5172        12.905201         77.686538
5173       5173        12.976893         77.576333
5174       5174        12.898712         77.750399

[5175 rows x 3 columns]

### Destination Data

In [14]:
# We create here new column for DESTINATION_ID values
df_des.insert(loc=0, column='DESTINATION_ID', value=range(0 , 0 + len(df_des)))

In [15]:
df_des.rename(columns={'Latitude' : 'DESTINATION_Latitude' , 'Longitude' : 'DESTINATION_Longitude'} , inplace=True)

In [16]:
df_des

DESTINATION_ID  DESTINATION_Latitude  DESTINATION_Longitude
0                  0             28.519427              76.964814
1                  1             28.521104              76.996469
2                  2             28.522902              76.910281
3                  3             28.523170              76.986546
4                  4             28.526302              76.953613
...              ...                   ...                    ...
4812            4812             28.715997              77.268173
4813            4813             28.716010              77.117582
4814            4814             28.716233              77.113719
4815            4815             28.716576              77.267506
4816            4816             28.716757              77.112978

[4817 rows x 3 columns]

### Calculating distance between source and destination points with formula

In [17]:

from datetime import datetime
start_time = datetime.now()


from math import sin, cos, sqrt, atan2, radians, pi


lst=[]
r1=[]
r2=[]
r3=[]
q1=[]
q2=[]
q3=[]

for r in df_src.itertuples():
    for q in df_des.itertuples():
        
        #R = 6371.0         # approximate radius of earth in km

        lat1 = pi/180 * (r[2])
        lon1 = pi/180 * (r[3])
        lat2 = pi/180 * (q[2])
        lon2 = pi/180 * (q[3])
            
        #dlon = lon2 - lon1
        #dlat = lat2 - lat1

        #a = (sin((lat2 - lat1)/2)**2+cos(lat1)*cos(lat2)*sin((lon2 - lon1)/2)**2)
        #c = 2 * atan2(sqrt(a), sqrt(1 - a))

        dis = 6371 * ( 2 * atan2( ((sin((lat2 - lat1)/2)**2+cos(lat1)*cos(lat2)*sin((lon2 - lon1)/2)**2))**0.5 , 
                                 (1 - (sin((lat2 - lat1)/2)**2+cos(lat1)*cos(lat2)*sin((lon2 - lon1)/2)**2))**0.5 ) )
        
        lst.append(dis)
        r1.append(r[1])
        r2.append(r[2])
        r3.append(r[3])
        q1.append(q[1])
        q2.append(q[2])
        q3.append(q[3])
        
            

        
    
    
end_time = datetime.now()
print('Duration: {}'.format(end_time - start_time))

Duration: 0:02:33.316228


In [18]:
from datetime import datetime
start_time = datetime.now()

r11 = pd.DataFrame(r1, columns =['SOURCE_ID'])
r11['SOURCE_ID'] = r11['SOURCE_ID'].astype(int)
r22 = pd.DataFrame(r2, columns =['SOURCE_Latitude'])
r33 = pd.DataFrame(r3, columns =['SOURCE_Longitude'])

q11 = pd.DataFrame(q1, columns =['DESTINATION_ID'])
q11['DESTINATION_ID'] = q11['DESTINATION_ID'].astype(int)
q22 = pd.DataFrame(q2, columns =['DESTINATION_Latitude'])
q33 = pd.DataFrame(q3, columns =['DESTINATION_Longitude'])

result = pd.DataFrame(lst, columns =['DISTANCE_KM'])
result['DISTANCE_KM'] = result['DISTANCE_KM'].astype(float)

df = pd.concat([r11,r22,r33,q11,q22,q33,result], axis=1)

    
end_time = datetime.now()
print('Duration: {}'.format(end_time - start_time))

Duration: 0:00:39.230532


In [19]:
# Final DataFrame with all calculations
df

SOURCE_ID  SOURCE_Latitude  SOURCE_Longitude  DESTINATION_ID  \
0                 0        13.022193         77.697762               0   
1                 0        13.022193         77.697762               1   
2                 0        13.022193         77.697762               2   
3                 0        13.022193         77.697762               3   
4                 0        13.022193         77.697762               4   
...             ...              ...               ...             ...   
24927970       5174        12.898712         77.750399            4812   
24927971       5174        12.898712         77.750399            4813   
24927972       5174        12.898712         77.750399            4814   
24927973       5174        12.898712         77.750399            4815   
24927974       5174        12.898712         77.750399            4816   

          DESTINATION_Latitude  DESTINATION_Longitude  DISTANCE_KM  
0                    28.519427              76.964814  1724.883195  
1                    28.521104              76.996469  1724.928447  
2                    28.522902              76.910281  1725.526640  
3                    28.523170              76.986546  1725.201427  
4                    28.526302              76.953613  1725.698226  
...                        ...                    ...          ...  
24927970             28.715997              77.268173  1759.509488  
24927971             28.716010              77.117582  1760.021641  
24927972             28.716233              77.113719  1760.061301  
24927973             28.716576              77.267506  1759.575810  
24927974             28.716757              77.112978  1760.122424  

[24927975 rows x 7 columns]

In [20]:
# End Time
end_time1 = datetime.now()
print('Overall Total Execution time: {}'.format(end_time1 - start_time1))

Overall Total Execution time: 0:03:16.313905


# Conclusion :
## - Here we calculated all combinations of distances 5175*4817=2.492 crore combinations.
## - Here we optimised code with using 'for loop', itertuples, lists, math library for calculations with formula to reduce execution time for such larger number of calculations.
## - We calculated distances and created dataframe with minimum time.
## - Note: Here time will depend on your individual computer specifications and performance, in my case it have done all tasks one in around 3-4 mins.